Adopt dynamicDowncast<> in a number of page classes

annevk opened the pull request:
#### b0ccbb24dcd4e4d30ac3c5b617daafc578d9cac8
<pre>
Adopt dynamicDowncast&lt;&gt; in a number of page classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=267512">https://bugs.webkit.org/show_bug.cgi?id=267512</a>

Reviewed by Darin Adler.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::searchForLabelsBeforeElement):
(WebCore::LocalFrame::setPageAndTextZoomFactors):
(WebCore::LocalFrame::contentFrameFromWindowOrFrameElement):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::init):
(WebCore::LocalFrameView::applyOverflowToViewport):
(WebCore::LocalFrameView::applyPaginationToViewport):
(WebCore::LocalFrameView::addEmbeddedObjectToUpdate):
(WebCore::LocalFrameView::canShowNonOverlayScrollbars const):
(WebCore::LocalFrameView::maintainScrollPositionAtAnchor):
(WebCore::countRenderedCharactersInRenderObjectWithThreshold):
(WebCore::LocalFrameView::updateEmbeddedObject):
(WebCore::LocalFrameView::hasCustomScrollbars const):
(WebCore::LocalFrameView::addChild):
(WebCore::LocalFrameView::removeChild):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::editableElementsInRect const):
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::isValidSampleLocation):
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::PageConsoleClient::screenshot):
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::PageSerializer::SerializerMarkupAccumulator::appendCustomAttributes):
(WebCore::PageSerializer::serializeFrame):
(WebCore::PageSerializer::serializeCSSStyleSheet):
(WebCore::PageSerializer::retrieveResourcesForProperties):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDispatchSimulatedMouseEvents const):
(WebCore::Quirks::shouldDispatchedSimulatedMouseEventsAssumeDefaultPrevented const):
(WebCore::Quirks::simulatedMouseEventTypeForTarget const):
(WebCore::Quirks::shouldMakeTouchEventNonCancelableForTarget const):
(WebCore::Quirks::shouldPreventDispatchOfTouchEvent const):
(WebCore::Quirks::shouldBypassBackForwardCache const):
(WebCore::Quirks::shouldMakeEventListenerPassive):
(WebCore::Quirks::triggerOptionalStorageAccessQuirk const):

Canonical link: <a href="https://commits.webkit.org/273154@main">https://commits.webkit.org/273154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a685f7884518f1f5733735637984a707e690ab1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34448 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13267 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36450 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37135 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31154 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35537 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15657 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10377 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30148 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34969 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30716 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9809 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9894 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30771 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38414 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31287 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31086 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35964 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10014 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7899 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33908 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11829 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7925 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10570 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10872 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -496,11 +496,9 @@ bool Quirks::shouldDispatchSimulatedMouseEvents(const EventTarget* target) const
-                    return true;
-            }
+        for (RefPtr node = dynamicDowncast<Node>(target); node; node = node->parentNode()) {
+            if (RefPtr element = dynamicDowncast<Element>(node); element && const_cast<Element&>(element).classList().contains("lm-DockPanel-tabBar"_s))
```
I also tried this without const_cast<Element&>() and that didn't compile either on Linux. @mcatanzaro do you perhaps have advice for me? Thanks!

mcatanzaro replied:
So you have a `RefPtr<const Element>` (unfortunate!) and you want to get a non-const `Element` from it? You're surely not going to be able to cast from `RefPtr` directly to `Element` (how does that work? really surprised that builds on Apple platforms). Instead, try getting at the `const Element` first, _then_ do the `const_cast`. I would try: `const_cast<Element*>(element.get())->classList()`

annevk replied:
Thank you for identifying the issue so clearly. Your suggestion didn't work, but removing `const` altogether seems to work fine and it wasn't really buying us anything here.

annevk commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -564,9 +565,9 @@ bool Quirks::shouldMakeTouchEventNonCancelableForTarget(EventTarget* target) con
     auto host = m_document->topDocument().url().host();
 
     if (equalLettersIgnoringASCIICase(host, "www.youtube.com"_s)) {
```
This needs to be a follow-up PR, but this makes no sense. _host_ is canonicalized by the URL parser so we can compare for equality.

darinadler replied:
If by "makes no sense" you mean "is slightly suboptimal", then agreed.

cdumez commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2281,7 +2281,7 @@ void Node::moveNodeToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-bool isTouchRelatedEventType(const EventTypeInfo& eventType, const EventTarget& target)
+bool isTouchRelatedEventType(const EventTypeInfo& eventType, EventTarget& target)
```
Dropping const here seems odd. The function doesn't seem like something that should need to modify the eventTarget. Odds are we want to add an `Element::classList()` getter that is const.

darinadler commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -500,11 +500,14 @@ String LocalFrame::searchForLabelsBeforeElement(const Vector<String>& labels, El
 
         if (n->hasTagName(tdTag) && !startingTableCell)
-            startingTableCell = downcast<HTMLTableCellElement>(n);
+            startingTableCell = uncheckedDowncast<HTMLTableCellElement>(n);
```
I think it would be safer to just write `dynamicDowncast` here.

cdumez replied:
Why not just keep downcast() since the expectation is that this will be the type?

annevk replied:
I guess there's something I don't understand. Given the tdTag check, why would this ever fail?

darinadler replied:
It wouldn’t but the whole point is to have code that can be verified locally without understanding connections like the one between a tag and a DOM class. We aren’t trying to make “carefully written” code where we make no mistakes but rather obviously safe code.

annevk replied:
I see, and so only when it's a hot path would we consider using uncheckedDowncast? I should revert a change to uncheckedDowncast in another PR as well then.

cdumez replied:
The check shouldn't fail but it's not necessarily obvious. I would also prefer to stay on the safe side and keep downcast<>() in this particular instance. Better safe than sorry.

annevk commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -496,11 +496,9 @@ bool Quirks::shouldDispatchSimulatedMouseEvents(const EventTarget* target) const
-                    return true;
-            }
+        for (RefPtr node = dynamicDowncast<Node>(target); node; node = node->parentNode()) {
+            if (RefPtr element = dynamicDowncast<Element>(node); element && const_cast<Element&>(*element).classList().contains("lm-DockPanel-tabBar"_s))
```
@cdumez I suspect that if I make this `auto*` again it'll work for GCC. Thoughts?

annevk committed:
Adopt dynamicDowncast<> in a number of page classes
https://bugs.webkit.org/show_bug.cgi?id=267512

Reviewed by Darin Adler.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::searchForLabelsBeforeElement):
(WebCore::LocalFrame::setPageAndTextZoomFactors):
(WebCore::LocalFrame::contentFrameFromWindowOrFrameElement):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::init):
(WebCore::LocalFrameView::applyOverflowToViewport):
(WebCore::LocalFrameView::applyPaginationToViewport):
(WebCore::LocalFrameView::addEmbeddedObjectToUpdate):
(WebCore::LocalFrameView::canShowNonOverlayScrollbars const):
(WebCore::LocalFrameView::maintainScrollPositionAtAnchor):
(WebCore::countRenderedCharactersInRenderObjectWithThreshold):
(WebCore::LocalFrameView::updateEmbeddedObject):
(WebCore::LocalFrameView::hasCustomScrollbars const):
(WebCore::LocalFrameView::addChild):
(WebCore::LocalFrameView::removeChild):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::editableElementsInRect const):
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::isValidSampleLocation):
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::PageConsoleClient::screenshot):
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::PageSerializer::SerializerMarkupAccumulator::appendCustomAttributes):
(WebCore::PageSerializer::serializeFrame):
(WebCore::PageSerializer::serializeCSSStyleSheet):
(WebCore::PageSerializer::retrieveResourcesForProperties):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDispatchSimulatedMouseEvents const):
(WebCore::Quirks::shouldDispatchedSimulatedMouseEventsAssumeDefaultPrevented const):
(WebCore::Quirks::simulatedMouseEventTypeForTarget const):
(WebCore::Quirks::shouldMakeTouchEventNonCancelableForTarget const):
(WebCore::Quirks::shouldPreventDispatchOfTouchEvent const):
(WebCore::Quirks::shouldBypassBackForwardCache const):
(WebCore::Quirks::shouldMakeEventListenerPassive):
(WebCore::Quirks::triggerOptionalStorageAccessQuirk const):

Canonical link: https://commits.webkit.org/273154@main
